File: src/main/java/edu/cornell/cs/apl/viaduct/imp/transformers/Elaborator.java
package edu.cornell.cs.apl.viaduct.imp.transformers;

import edu.cornell.cs.apl.viaduct.imp.ast.BlockNode;
import edu.cornell.cs.apl.viaduct.imp.ast.BreakNode;
import edu.cornell.cs.apl.viaduct.imp.ast.ExpressionNode;
import edu.cornell.cs.apl.viaduct.imp.ast.ForNode;
import edu.cornell.cs.apl.viaduct.imp.ast.IfNode;
import edu.cornell.cs.apl.viaduct.imp.ast.LoopNode;
import edu.cornell.cs.apl.viaduct.imp.ast.ProgramNode;
import edu.cornell.cs.apl.viaduct.imp.ast.StatementNode;
import edu.cornell.cs.apl.viaduct.imp.ast.WhileNode;
import edu.cornell.cs.apl.viaduct.imp.visitors.IdentityProgramVisitor;
import edu.cornell.cs.apl.viaduct.imp.visitors.StmtVisitor;

/** Elaborate derived AST nodes into internal ones. */
public class Elaborator {
  public static ProgramNode run(ProgramNode program) {
    return new ElaborationProgramVisitor().run(program);
  }

  // TODO: we shouldn't need this
  public static StatementNode run(StatementNode statement) {
    return new ElaborationProgramVisitor().run(statement);
  }

  private static class ElaborationProgramVisitor extends IdentityProgramVisitor {
    private final StmtVisitor<StatementNode> statementVisitor = new ElaborateStmtVisitor();

    @Override
    protected StmtVisitor<StatementNode> getStatementVisitor() {
      return statementVisitor;
    }

    protected class ElaborateStmtVisitor extends IdentityStmtVisitor {
      /** Rewrite while loops into loop-until-break loops. */
      @Override
      protected StatementNode leave(
          WhileNode node, IdentityStmtVisitor visitor, ExpressionNode guard, StatementNode body) {

        /*
         * while (guard) { body... }
         *   gets translated to
         * loop {
         *   if (guard) {
         *     body...
         *   } else {
         *     break;
         *   }
         * }
         */
        final BreakNode breakNode = BreakNode.builder().setSourceLocation(node).build();
        final IfNode resultGuard =
            IfNode.builder()
                .setGuard(guard)
                .setThenBranch((BlockNode) body)
                .setElseBranch(BlockNode.builder().add(breakNode).setSourceLocation(node).build())
                .setSourceLocation(node)
                .build();
        return LoopNode.builder()
            .setBody(BlockNode.builder().add(resultGuard).build())
            .setSourceLocation(node)
            .build();
      }

      @Override
      protected StatementNode leave(
          ForNode node,
          IdentityStmtVisitor visitor,
          StatementNode initialize,
          ExpressionNode guard,
          StatementNode update,
          StatementNode body) {

        /*
         * for (init; guard; update) { body... }
         *   gets translated to
         * {
         *   init;
         *   loop {
         *     if (guard) {
         *       body...
         *       update;
         *     } else {
         *       break;
         *     }
         *   }
         * }
         */

        final BlockNode.Builder result = BlockNode.builder().setSourceLocation(node);

        result.add(initialize);

        final BreakNode breakNode = BreakNode.builder().setSourceLocation(node).build();
        final BlockNode resultBody = ((BlockNode) body).toBuilder().add(update).build();
        final IfNode resultLoopBody =
            IfNode.builder()
                .setGuard(guard)
                .setThenBranch(resultBody)
                .setElseBranch(BlockNode.builder().add(breakNode).setSourceLocation(node).build())
                .setSourceLocation(node)
                .build();

        result.add(
            LoopNode.builder()
                .setBody(BlockNode.builder().add(resultLoopBody).setSourceLocation(node).build())
                .setSourceLocation(node)
                .build());

        return result.build();
      }
    }
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/syntax/transformers/Elaborator.kt
package edu.cornell.cs.apl.viaduct.syntax.transformers

import edu.cornell.cs.apl.viaduct.syntax.surface.BlockNode
import edu.cornell.cs.apl.viaduct.syntax.surface.BreakNode
import edu.cornell.cs.apl.viaduct.syntax.surface.ForLoopNode
import edu.cornell.cs.apl.viaduct.syntax.surface.IfNode
import edu.cornell.cs.apl.viaduct.syntax.surface.InfiniteLoopNode
import edu.cornell.cs.apl.viaduct.syntax.surface.StatementNode
import edu.cornell.cs.apl.viaduct.syntax.surface.WhileLoopNode
import kotlinx.collections.immutable.toPersistentList

// TODO: this is broken AF. Doesn't recurse.
class Elaborator {
    companion object {
        fun run(stmt: StatementNode): StatementNode {
            return Elaborator().elaborate(stmt)
        }
    }

    /**
     * elaborate derived forms in a statement.
     *
     * @param stmt the statement to elaborate.
     */
    private fun elaborate(stmt: StatementNode): StatementNode {
        return when (stmt) {
            is WhileLoopNode -> {
                InfiniteLoopNode(
                    BlockNode(
                        IfNode(
                            stmt.guard,
                            stmt.body,
                            BlockNode(
                                BreakNode(sourceLocation = stmt.sourceLocation),
                                sourceLocation = stmt.sourceLocation
                            ),
                            stmt.sourceLocation
                        ), sourceLocation = stmt.sourceLocation
                    ),
                    null,
                    stmt.sourceLocation
                )
            }

            is ForLoopNode -> {
                val bodyBlock = mutableListOf<StatementNode>()
                bodyBlock.addAll(stmt.body.statements)
                bodyBlock.add(stmt.update)

                val newBody = BlockNode(bodyBlock.toPersistentList(), stmt.body.sourceLocation)

                val block = mutableListOf<StatementNode>()
                block.add(stmt.initialize)
                block.add(
                    InfiniteLoopNode(
                        BlockNode(
                            IfNode(
                                stmt.guard,
                                newBody,
                                BlockNode(
                                    BreakNode(null, stmt.sourceLocation),
                                    sourceLocation = stmt.sourceLocation
                                ),
                                stmt.guard.sourceLocation
                            ), sourceLocation = stmt.sourceLocation
                        ),
                        null,
                        stmt.sourceLocation
                    )
                )
                BlockNode(block.toPersistentList(), stmt.sourceLocation)
            }

            else -> stmt
        }
    }
}
